### DISCUSSION TITLE
How to count results from a query

### DISCUSSION THREAD

**Discussion by andregoldstein** [supabase/supabase#517] [Questions] [answered]
Hey guys I'm perhaps missing something simple here but how can I go about counting the results of a query?

I can grab a range of documents from a query using `range(0,10)` but I'd also like/need to be able to know how many total results matched in order to properly display pagination.

Thanks!

**Comment by soedirgo** [comment, marked as answer]
Hey @andregoldstein! There's a PR [here](https://github.com/supabase/postgrest-js/pull/147) that adds this feature. I'll be updating here when we update the client (should be sometime soon).

**Reply by andregoldstein** [reply]
Great! Thanks 👍

**Reply by soedirgo** [reply]
OK, the client has been updated, so now you should be able to do this:

```js
const { error, data, count } = await supabase
    .from('table')
    .select('*', { count: 'exact' })
```

Let me know if it works!

**Reply by andregoldstein** [reply]
Thanks! Will give it a go as soon as I can!

**Reply by himeshp** [reply]
<img width="504" alt="Screenshot 2021-07-24 at 3 56 57 PM" src="https://user-images.githubusercontent.com/5878422/126865692-bb9d119b-b6b3-4ba4-891f-604d6f24be79.png">
This does not seem to work on the dart library?

**Reply by soedirgo** [reply]
@himeshp for Dart you can try it like the "query data" example here: https://pub.dev/packages/supabase/example

**Reply by himeshp** [reply]
yes, thanks

**Comment by saltcod** [comment]
Hummm I think I might be missing something. I just want the count of a query: 

This works, but I don't think it's efficient?
```
let { data, error, status } = await supabase
	.from("users")
	.select("*", { count: "exact" }) // if you don't want to return any rows, you can use { count: 'exact', head: true }
	.eq("role", "developer");
	console.log(data.length);
```

Seem like this is the recommended solution, but doesn't work for me (returns null):
```
let { data, error, status } = await supabase
	.from("users")
	.select("*", { count: "exact", head: true })
	.eq("role", "developer");
	console.log(data); // null
```


Any nudges appreciated!

**Reply by GaryAustin1** [reply]
Count comes back in count.   Head returns no data so null.

**Reply by saltcod** [reply]
Right right.  { data, error, status, count }

duh! thanks for this!

**Reply by ogrotten** [reply]
I'm getting null from `count` here as well... so how is this done to get just a row count and no data?

I have: 

	const { data, error, status, count } = await supabase
		.from('users`)
		.select('*', { count: 'exact', head: true })
	console.log(count) // null
	
What am I missing here?

Is there a doc on it that I'm missing? I'm going from [this doc page here](https://supabase.com/docs/reference/javascript/select) and it doesn't mention the `status` parameter.

**Reply by GaryAustin1** [reply]
@ogrotten except for your typo on users which I doubt is in your real code, I don't see anything wrong with your call.  
Count comes back 0 if RLS is on and user does not have access, BUT it is null if table does not exist (you are not trying to access auth.users are you as you can't from the API).  
Are you getting an error?
Status reflects the http status (200,404).

**Reply by max-programming** [reply]
@GaryAustin1 I have the same code but I also get `null`
Except when I remove the `head` property, I get the array

```ts
const { data } = await supabase
  .from("properties")
  .select("*", { count: "exact", head: true })
console.log(data) // null
```
```ts
const { data } = await supabase
  .from("properties")
  .select("*", { count: "exact" })
console.log(data) // [{ ... }, { ... }]
```

UPDATE: I thought the `data` property contained the count, But there is a `count` property
```ts
const { count } = await supabase
  .from("properties")
  .select("*", { count: "exact", head: true })
console.log(count) // 745
```

**Comment by michael-sumner** [comment]
The `count` property helped me with obtaining the row count, thanks @max-programming !

```js
const { count, error } = await supabase
  .from("properties")
  .select("*", { count: "exact", head: true })

if (error) throw error;

console.log(count) // 745

return count;
```

**Comment by tjmcdonough** [comment]
How can you do this via an API call instead of using the supabase libary?

**Reply by tjmcdonough** [reply]
Do not worry, found it

```
Request
  GET /rest/v1/{table-name}
    ?id=gt.10
    &order=id.asc
    &limit=10
  apikey ...supabase access key...
  Prefer count=exact
Response
  Content-Range: 0-9/100
```

**Comment by antonio-bluco** [comment]
This doesn't seem to be available when selecting from a rpc results, i.e. 

`supabase.rpc('my_function').select('*', {count: 'exact'})`

The typing system complains that 'count' does not exist, and the count returned is in fact always null. 
Any ideas on what to do here?

**Reply by elyobo** [reply]
rpc seems to require that you pass `{ count: 'exact' }` in the rpc call instead, e.g.

```js
supabase.rpc('my_function',  {count: 'exact'}).select('*')
```

The returned type from `.rpc().select()` is different too, not sure whether it should be or not.

**Comment by BorisBarzotto** [comment]
rpc count always return null. Some idea?, no problem with data array but count always null.
using:
@supabase/supabase-js": "^2.42.7"
@supabase/ssr": "0.5.1"

**Reply by GaryAustin1** [reply]
Do you have a table returning function?
https://docs.postgrest.org/en/v12/references/api/functions.html
